### ISSUE TITLE
Fail on no test file for PublishTestResults@2

### ISSUE THREAD

**Issue by derekendres** [microsoft/azure-pipelines-tasks#13275]
## Note  
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:  

https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Feature

**Enter Task Name**: PublishTestResults@2  

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines or TFS on-premises?
    
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 

https://dev.azure.com/osieng/engineering/_build/results?buildId=103053&view=logs&j=846ec57a-2bea-5833-7752-007160b95b9a&t=d1056ef1-5a88-52db-be24-b861f9682b72 

- Agent - Hosted or Private: 
    
    - If using Hosted agent, provide agent queue name:

        linux:
          imageName: ubuntu-latest
        windows:
          imageName: windows-latest

## Issue Description

When publishing a test result, if there is no test result file, this task gives a warning that is eaten by a success indication.  It should flag as a warning that it didn't actually publish the result and ideally there should be an option to fail if the test file doesn't exist.

### Task logs

Starting: Publish test results
==============================================================================
Task         : Publish Test Results
Description  : Publish test results to Azure Pipelines
Version      : 2.171.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/test/publish-test-results
==============================================================================
/usr/bin/dotnet --version
3.1.301
##[warning]No test result files matching **/test-*.xml were found.
Finishing: Publish test results

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

[Insert error from the logs here for a quick overview]

**Comment by csurfleet** [comment]
This would be incredibly useful, we're using bash scripts between each test task at the moment

**Comment by BarsikV** [comment]
My organization and I would appreciate this feature very much, too!

> 
> 
> This would be incredibly useful, we're using bash scripts between each test task at the moment

Would you mind sharing your script? I was thinking at doing something like that, but I am not very good at parsing files using a script...

**Comment by bneumann** [comment]
> This would be incredibly useful, we're using bash scripts between each test task at the moment
Yeah if I need to script my way around problems like that I really don't need these tasks at all. I prefer a fix for this :)

**Comment by navin22** [comment]
Adding @shailesh-sk

**Comment by xncbf** [comment]
Anybody working on this issue? This option is essential

**Comment by walliski** [comment]
Coming in yet half year later.

This would be really essential for the pipelines I am working on. In the current setup I want to do, I would not want to fail the build based on the exit status of the tests step, but continue and later use the published test results further down in the pipeline. This means that both this task, and the Powershell task that runs the tests is set to not fail the build on failure. To safeguard from the testrunner failing completely and not running any tests at all, I would need this task to fail in case it cannot find any test result files!

**Comment by jmhon08** [comment]
I would like to see this as well. @shailesh-sk the PR is waiting your review

**Comment by shailesh-sk** [comment]
@jmhon08 PR looks good to me, feel free to checkin.

**Comment by 0r2W9vXGArZ8e0CYGxwS** [comment]
A bit late to the party, but I'm looking forward to seeing this added as well.

**Comment by jmhon08** [comment]
@upadhyayap any idea when this will be merged?

**Comment by walliski** [comment]
@shailesh-sk @upadhyayap

Did we get anywhere with #14868 ?

**Comment by mathiash98** [comment]
This is an essential feature... It should at least make a warning if not fail completely at the "Publish Test Results task"

**Comment by LynxJ** [comment]
Would really like to see this feature implemented

**Comment by ColeSiegelTR** [comment]
Also looking for this, seems fundamental

**Comment by mathiash98** [comment]
We ended up using follow bash script tasks:
- We also experienced after some bad upgrades that the test result file was created, but contained no results. As the `dotnet test` could not find any unit tests to run.

We rely on running tests inside a docker compose as we also test with database. But if you purely rely on the VSTest task you can use the argument `minimumExpectedTests` https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/vstest?view=azure-devops#arguments to fail if there are no tests run.

```bash
- bash: |
    if [ $(test_results_host_folder)**/*.trx ] && [ $(grep -E "<UnitTestResult" $(test_results_host_folder)**/*.trx -c) -gt 0 ]; then
      echo "##vso[task.setVariable variable=TESTRESULTFOUND]true"
    fi
  displayName: Check if test results file & results >= 1
- script: |
      echo No test result found
      exit 1
  displayName: No test result found
  condition: ne(variables.TESTRESULTFOUND, 'true')
```

**Comment by ColeSiegelTR** [comment]
been another 4 months any plans to fix this?

**Comment by b055man** [comment]
> Coming in yet half year later.
> 
> This would be really essential for the pipelines I am working on. In the current setup I want to do, I would not want to fail the build based on the exit status of the tests step, but continue and later use the published test results further down in the pipeline. This means that both this task, and the Powershell task that runs the tests is set to not fail the build on failure. To safeguard from the testrunner failing completely and not running any tests at all, I would need this task to fail in case it cannot find any test result files!

The problem is the PublishTestResults succeeds not only if there are no tests but also if it fails to parse the test results, e.g. if a failed test case produces output that cannot be parsed. 
This's just happened in our environment when some test cases were failing but this was not captured by the pipeline.
To me this sounds like a bug, not an enhancement!
![image](https://user-images.githubusercontent.com/1776065/175007589-1c20987f-45f3-4ccd-a876-cc7fa8036895.png)

**Comment by Danon** [comment]
I think it would make more sense if the job would fail by default, and maybe add an option to explicitly silence it.

**Comment by walliski** [comment]
> This issue is stale because it has been open for 180 days with no activity. Remove the stale label or comment on the issue otherwise this will be closed in 5 days

**Comment by AurimasNav** [comment]
I'd love to see this fail by default if no test files are found.

**Comment by jamesstojan1** [comment]
This would be really helpful. The default assumption when publishing non existent files should be to error out.

**Comment by coffeebe4code** [comment]
nearly 3 years this has been opened?

Everyone going to be here for its 3rd birthday in July?

**Comment by abatishchev** [comment]
Bump.

**Comment by xncbf** [comment]
I got tired of waiting and found the solution myself.
The root cause I've encountered is that the failure signal appears to be ignored if there is another command on the next line that runs pytest.
I solved it by receiving the signal from the line immediately following the test and processing it directly, as follows.
```
pipenv run pytest --junitxml=junit/test-results.xml --cov=. --cov-report=html
EXIT_CODE=$?
if [ $EXIT_CODE -ne 0 ]; then
exit $EXIT_CODE
fi
```

**Comment by ChristianStadelmann** [comment]
This issue is not stale, but I guess the bot is…

**Comment by KevinPartingtonTR** [comment]
Microsoft folks, if this is something you don't plan to fix, you should close this issue as not planned so we know. Thanks!

**Comment by dts12263** [comment]
[like]  Dustin Sauriol reacted to your message:
________________________________

**Comment by riQQ** [comment]
This is available with version 2.228.0+ via setting the input `failTaskOnMissingResultsFile` to `true`. It was implemented by #18931 and is also [documented on learn.microsoft.com](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/publish-test-results-v2?view=azure-pipelines&tabs=trx%2Ctrxattachments%2Cyaml#syntax).

```yaml
- task: PublishTestResults@2
  inputs:
    failTaskOnMissingResultsFile: true # boolean. Fail if no result files are found. Default: false.
```

**Comment by coffeebe4code** [comment]
I'm in a lot of interesting github issues, and this is certainly one of them. We shouldnt need a variable to fail a task that should fail. If someone is relying on a test task passing when there are no tasks, they are almost certainly not aware of it, why not make them use a flag, pass on no tests.

Mind blowing this is where we are.

**Comment by abatishchev** [comment]
What do you mean "a variable"? A parameter? If yes, then why not? What would be other options?

Here are the parameters currently avaliable, which make sense to me:

```yaml
#failTaskOnFailedTests: false # boolean. Fail if there are test failures. Default: false.
#failTaskOnFailureToPublishResults: false # boolean. Fail if there is failure in publishing test results. Default: false.
#failTaskOnMissingResultsFile: false # boolean. Fail if no result files are found. Default: false.
```

Per the [doc](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/publish-test-results-v2).

**Comment by coffeebe4code** [comment]
1. Hey there is a bug in your pipeline job. It passes if there are no tests found. We had an issue where the test task was not ran in the right folder location, and it passed, even though we had tests failing.
2. Crazy! this isn't a bug though. We could add a parameter so it doesn't interrupt others. I updated the docs and implemented this.
3. .... no like it was a bug, just fail the task if there are no tests found.
4. Here, it says it in the docs you need to set this parameter to false.

...

**Comment by ChristianStadelmann** [comment]
Exactly. All of those boolean flags should default to `true`, which is the only way that the pipeline task roughly does what it is expected to do. Defaulting them to `false` is counterintuitive, hides failures, hiding bugs.
